tsdb/engine/wal/wal.go

```diff
@@ -0,0 +1,1596 @@
+		return err
+	}
+
+	return l.metaFile.Sync()
+}
+
+// nextMetaFile will close the current file if there is one open and open a new file to log
+// metadata updates to. This function assumes that you've locked l.mu elsewhere.
+func (l *Log) nextMetaFile() error {
+	if l.metaFile != nil {
+		if err := l.metaFile.Close(); err != nil {
+			return err
+		}
+	}
+
+	metaFiles, err := l.metadataFiles()
+	if err != nil {
+		return err
+	}
+
+	id := 0
+	if len(metaFiles) > 0 {
+		num := strings.Split(filepath.Base(metaFiles[len(metaFiles)-1]), ".")[0]
+		n, err := strconv.ParseInt(num, 10, 32)
+
+		if err != nil {
+			return err
+		}
+
+		id = int(n)
+	}
+
+	nextFileName := filepath.Join(l.path, fmt.Sprintf("%06d.%s", id, MetaFileExtension))
+	l.metaFile, err = os.OpenFile(nextFileName, os.O_CREATE|os.O_RDWR, 0666)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// metadataFiles returns the files in the WAL directory with the MetaFileExtension
+func (l *Log) metadataFiles() ([]string, error) {
+	path := filepath.Join(l.path, fmt.Sprintf("*.%s", MetaFileExtension))
+
+	a, err := filepath.Glob(path)
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Strings(a)
+
+	return a, nil
+}
+
+// pointsToPartitions returns a map that organizes the points into the partitions they should be mapped to
+func (l *Log) pointsToPartitions(points []tsdb.Point) map[*Partition][]tsdb.Point {
+	m := make(map[*Partition][]tsdb.Point)
+	for _, p := range points {
+		pp := l.partition(p.Key())
+		m[pp] = append(m[pp], p)
+	}
+	return m
+}
+
+// openPartitionFiles will open all partitions and read their segment files
+func (l *Log) openPartitionFiles() error {
+	results := make(chan error, len(l.partitions))
+	for _, p := range l.partitions {
+
+		go func(p *Partition) {
+			// Recover from a partial compaction.
+			if err := p.recoverCompactionFile(); err != nil {
+				results <- fmt.Errorf("recover compaction files: %s", err)
+				return
+			}
+
+			fileNames, err := p.segmentFileNames()
+			if err != nil {
+				results <- err
+				return
+			}
+			for _, n := range fileNames {
+				entries, err := p.readFile(n)
+				if err != nil {
+					results <- err
+					return
+				}
+				for _, e := range entries {
+					p.addToCache(e.key, e.data, e.timestamp)
+				}
+			}
+			results <- nil
+		}(p)
+	}
+
+	for i := 0; i < len(l.partitions); i++ {
+		err := <-results
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Close will finish any flush that is currently in process and close file handles
+func (l *Log) Close() error {
+	// stop the autoflushing process so it doesn't try to kick another one off
+	if l.closing != nil {
+		close(l.closing)
+		l.closing = nil
+	}
+
+	l.wg.Wait()
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	// clear the cache
+	l.partitions = nil
+
+	return l.close()
+}
+
+// close all the open Log partitions and file handles
+func (l *Log) close() error {
+	for _, p := range l.partitions {
+		if err := p.Close(); err != nil {
+			return err
+		}
```

Comment:
If one of these errors out then the other partitions won't be closed. I would skip the error.